// raft.proto
syntax = "proto3";
package raft;

option java_package = "com.example.raft";
option go_package = "github.com/example/distributed-ticket-booking/proto/raft;raft";

// Minimal Raft RPCs to implement AppendEntries and RequestVote.
// These messages are intentionally explicit to support log replication, heartbeats,
// and leader election.

service Raft {
  // Leader -> follower: replicate log entries / heartbeat
  rpc AppendEntries(AppendEntriesRequest) returns (AppendEntriesResponse);

  // Candidate -> peer: request votes during election
  rpc RequestVote(RequestVoteRequest) returns (RequestVoteResponse);
}

message LogEntry {
  int64 index = 1;
  int64 term = 2;
  string command = 3;        // serialized command (e.g., JSON or protobuf bytes as string/base64)
  bytes data = 4;            // optional raw payload
}

message AppendEntriesRequest {
  int64 term = 1;                // leader term
  string leader_id = 2;          // leader identity (address or id)
  int64 prev_log_index = 3;      // index of log entry immediately preceding new ones
  int64 prev_log_term = 4;       // term of prev_log_index entry
  repeated LogEntry entries = 5;// log entries to store (empty for heartbeat)
  int64 leader_commit = 6;       // leader's commit index
}

message AppendEntriesResponse {
  int64 term = 1;               // currentTerm, for leader to update itself
  bool success = 2;             // true if follower contained entry matching prev_log_index/term
  int64 match_index = 3;        // highest log index known to be replicated on follower (optional)
  string message = 4;           // for debugging / errors
}

message RequestVoteRequest {
  int64 term = 1;               // candidate's term
  string candidate_id = 2;      // candidate requesting vote
  int64 last_log_index = 3;     // index of candidate's last log entry
  int64 last_log_term = 4;      // term of candidate's last log entry
}

message RequestVoteResponse {
  int64 term = 1;               // currentTerm, for candidate to update itself
  bool vote_granted = 2;        // true means candidate received vote
  string message = 3;
}
